### ISSUE TITLE
Add setting to enable strict checking of non-strict Pydantic models

### ISSUE THREAD

**Issue by tibbe** [astral-sh/ty#4540]
### Summary

ty models Pydantic constructors with lax coercion, so both of these pass:

```py
import pydantic

class Model(pydantic.BaseModel):
    count: int

Model(count='123')  # ty: ok — coerces to 123 at runtime
Model(count='abc')  # ty: ok — but raises ValidationError at runtime
```

The first is correct by the book. The second is the problem: ty accepts a call that always fails at runtime, because `str` is a valid lax input for `int` in general but this particular `str` is not.

`pydantic.mypy` with `init_typed = True` rejects both:

```
error: Argument "count" to "Model" has incompatible type "str"; expected "int"  [arg-type]
```

Teams that set `init_typed = True` have deliberately opted out of relying on coercion — they want declared field types enforced at construction. There is currently no way to express that in ty, which blocks migrating any codebase using this setting.

Request: an opt-in rule, off by default, that checks constructor arguments against the declared field types rather than the lax input types.

This is the same shape as #3966, where ty correctly honored `extra=ignore` but teams wanted the stricter check regardless, and the resolution was a rule they could turn off. The default would be reversed here: off by default, enabled by teams coming from `init_typed = True`.

### Version

ty 0.0.81

**Comment by sharkdp** [comment]
Thank you for the report.

Did you see that ty validates constructor signatures in a strict way when your models use Pydantic's `strict=True` validation mode? Is that an option for you here? (why not?)

**Comment by tibbe** [comment]
> Thank you for the report.
> 
> Did you see that ty validates constructor signatures in a strict way when your models use Pydantic's `strict=True` validation mode? Is that an option for you here? (why not?)

It does work.

The catch is that it couples two migrations. `init_typed` is there to catch a
cheap programmer error — passing a wrong typed value into a constructor — and
we'd like to keep that check when we move to ty. Getting it via `strict=True`
also changes runtime validation, which we do rely on: numbers come back from
DynamoDB through boto3 as `Decimal`, and lax coercion turns them into `int`.
Keeping that means adding `strict=False` at every place we deserialize — around
245 call sites — and FastAPI validates request bodies itself, so the flag can't
be passed there at all. A call site we miss fails at runtime rather than in the
checker.

We may well move to strict eventually. We'd just rather not take that risk as
part of switching type checkers.

So not a must-have, but a static-only option would give teams already using
`init_typed` a low-risk path onto ty. #3966 looked like a similar case: accurate
default, stricter opt-in on top.

**Comment by tibbe** [comment]
I realize our code base isn't in an "ideal" state and it would probably be better if we had used `strict=True` from day one (assuming it was an option when we started using pydantic). The realities of a large code base is just that a runtime behavior change is risky at this size.

**Comment by sharkdp** [comment]
> Teams that set `init_typed = True` have deliberately opted out of relying on coercion — they want declared field types enforced at construction.

> Getting it via `strict=True`
> also changes runtime validation, which we do rely on: numbers come back from
> DynamoDB through boto3 as `Decimal`, and lax coercion turns them into `int`.

These two statements seem to contradict each other? Wouldn't mypy's `init_typed` flag those `Decimal` -> `int` coercions? Or does this particular coercion not happen in a constructor call (using `MyModel.model_validate(..)` or similar)?

> So not a must-have, but a static-only option would give teams already using `init_typed` a low-risk path onto ty. [#3966](https://github.com/astral-sh/ty/issues/3966) looked like a similar case: accurate default, stricter opt-in on top.

Implementing this using a new opt-in diagnostic seems doable, but would need some design work / prototyping. Enabling or disabling a rule should not change type inference. This means that we would probably keep the `count` parameter typed as `LaxInt` if that rule is enabled, but somehow additionally check the argument against the strict `int` type. If it the argument is assignable to `LaxInt`, but not to `int`, we emit the opt-in diagnostic. But there are lots of questions:

- how would this work for field types like `list[int]` (which are mapped to `Iterable[LaxInt]` for lax-mode models)?
- how would this interact with generics?
- how would this interact with bidirectional inference?
- how would this work work with custom validators?

**Comment by tibbe** [comment]
> These two statements seem to contradict each other? Wouldn't mypy's `init_typed` flag those `Decimal` -> `int` coercions? Or does this particular coercion not happen in a constructor call (using `MyModel.model_validate(..)` or similar)?

Exactly. On the parse from DB path we use generic code that relies on the lax coercion (e.g. to handle `Decimal`). So in our codebase the `init_typed` flag guards against incorrectly calling the constructor explicitly.

(I'm separately trying to convert our codebase to `strict=True` and it unfortunately causes lots of failures in e.g. unit tests if you just switch it on, adding weight to the argument above that for a sufficiently large codebase just switching `init_typed` -> `strict=True` isn't trivial.)

**Comment by sharkdp** [comment]
Ok, thank your for the explanation. I see that this is a valid use case, but I'd love to get some more feedback from others before we add a new configuration option to ty. I've renamed the issue so better reflect the intended solution.
